Make VPA and Checkpoint updates concurrent

voelzmo opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
Introduce a worker pool for updates to VPA status and VPACheckpoints, such that vpa-recommender can deal with large-scale systems. In the current implementation updates are executed sequentially, which means that at some point the latency from vpa-recommender to kubernetes-apiserver will become the limiting factor for updates per second.

This PR introduces a new parameter, `--update-worker-count`, defaulting to `10`, which determines the size of the worker pool. This can be used to control the load on the kubernetes-apiserver.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Closes #7951
Inspired by https://github.com/kubernetes/autoscaler/compare/master...tkukushkin:autoscaler:parallel-updates
Thanks @tkukushkin!


#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
BREAKING: deprecated recommender flag `min-checkpoints`, progress is now ensured by the new `update-worker-count` flag. This flag will be removed in a later release.
FEATURE: Make updates to VPA status and VPACheckpoints concurrent to perform better in clusters with a large amount of VPAs.
Added new flag `update-worker-count` to the recommender to control the amount of concurrent workers for VPA and VPACheckpoint updates.
Increased defaults for client-side rate limits to `kube-api-qps=50.0` and `kube-api-burst=100.0` such that the settings make sense together with the default `update-worker-count=10`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/checkpoint/checkpoint_writer.go:
```diff
@@ -76,8 +77,54 @@ func getVpasToCheckpoint(clusterVpas map[model.VpaID]*model.Vpa) []*model.Vpa {
+	var wg sync.WaitGroup
+
+	// Start workers
+	for i := 0; i < concurrentWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for vpaToCheckpoint := range vpaCheckpointUpdates {
+				processCheckpointUpdateForVPA(vpaToCheckpoint, writer)
+			}
+		}()
+	}
```
I'm not sure I understand, can you please explain the logic here?
let's say I have 100 vpas and 10 workers. then according to your code each worker will go over all the vpas. what am I missing?

voelzmo replied:
`vpaCheckpointUpdates` is a channel where the VPAs that should be checkpointed are passed in. One of the workers will receive the VPA and processes it, all the others don't receive the message.

omerap12 replied:
Oh I understand. thanks!

omerap12 reviewed (commented):
I had actually started implementing this too, but you were quicker! I took a different approach, so I'd love to dive deeper into what you did.

voelzmo commented:
> I had actually started implementing this too, but you were quicker! I took a different approach, so I'd love to dive deeper into what you did.

Oh, sorry, I didn't realize you assigned the issue to yourself and that I didn't do this ;( Apologies for the double work.

omerap12 commented:
> > I had actually started implementing this too, but you were quicker! I took a different approach, so I'd love to dive deeper into what you did.
> 
> Oh, sorry, I didn't realize you assigned the issue to yourself and that I didn't do this ;( Apologies for the double work.

No biggie your work will be better anyway ;)

omerap12 commented on vertical-pod-autoscaler/pkg/recommender/checkpoint/checkpoint_writer.go:
```diff
@@ -76,8 +77,54 @@ func getVpasToCheckpoint(clusterVpas map[model.VpaID]*model.Vpa) []*model.Vpa {
 	vpas := getVpasToCheckpoint(writer.cluster.VPAs())
+
+	// Create a channel to send VPA updates to workers
+	vpaCheckpointUpdates := make(chan *model.Vpa)
```
Another question, is there any reason no to use a buffered channel?

voelzmo replied:
I'm not sure, tbh. I think I experimented with using buffered channels and didn't reach the point where it makes much of a difference.
But we could also switch back to buffered channels – the workers keep reading from the channel until the buffer is empty, even when the sender already closed it.

bboreham commented:
PMFBI but I expected some note here about the client-side rate limit.
I see it came up on the issue #7951: suggest to mention here in the release note or documentation that this limit must be increased before you see any benefit from concurrency.

voelzmo commented:
/hold

Still some todos on this:
* adjust default values for client-side rate limiting to make sense together with the defaults for concurrent workers
* add comment on the concurrent workers flag to mention that also client-side rate limit needs to be considered as per @bboreham's suggestion 
* bring back buffered channels as per @omerap12's suggestion

jackfrancis commented on vertical-pod-autoscaler/pkg/recommender/checkpoint/checkpoint_writer.go:
```diff
@@ -76,8 +77,54 @@ func getVpasToCheckpoint(clusterVpas map[model.VpaID]*model.Vpa) []*model.Vpa {
+	}
+}
+
+func (writer *checkpointWriter) StoreCheckpoints(ctx context.Context, minCheckpoints int, concurrentWorkers int) error {
```
I would love to see a UT where `minCheckpoints` are fewer than `concurrentWorkers` to prove out what happens when we return early from this func (`if ctx.Err() != nil && minCheckpoints <= 0 {`) before we close all the channens (`close(vpaCheckpointUpdates)`)

voelzmo replied:
You're right, after writing a test case I think instead of doing 
```
		if ctx.Err() != nil && minCheckpoints <= 0 {
			return ctx.Err()
		}
```

this should rather be a `break` to properly close the channel and wait for the currently queued vpas to be processed rather than killing the workers right in the middle of processing.

When we're doing it like that, I'm thinking that we probably don't even need an additional `minCheckpoints` flag anymore? This would rather be determined by the buffer size of the channel, which controls how many vpa objects we would enqueue (and wait for them to be finished).

We wouldn't have a direct count for the minimum amount of checkpoints in each loop, because each VPA could have a different amount, but probably it is comparable in the sense that it guarantees _some_ progress is made during checkpointing in each loop.

In turn, this also means that we probably don't want to set the buffer size of the channel to `len(vpas)`, but rather to e.g. `concurrentWorkers`?

WDYT?

jackfrancis replied:
I like that approach. Also let's get rid of this which I spent some time and can't figure out why we want it:

```
		select {
		case <-ctx.Done():
		default:
		}
```

I played around and it doesn't seem that there's any reason for the above:

https://play.golang.com/p/_4RMak0bBCE

I would expect your propose changes should not upset existing UT, so maybe ship the `minCheckpoints` refactor without touching existing cases so we can evaluate if any outcomes changed.

voelzmo replied:
Thanks for the testing, I was also kind of confused if `ctx.Err()` only gets filled after `ctx.Done()` has been called. However, I ended up switching to the pattern I'm used to the most: using
```
select {
case <- ctx.Done()
```
to stop the current routine

Let me know if that's what you had in mind or if you prefer a different way of doing things.

jackfrancis replied:
I know this has changed a lot since this thread begun, just wanted to weigh in and things currently look good!

adrianmoisey commented:
The release note section of the PR description has a typo:

> Increased defaults for client-side rate limits to `kube-api-qps=50.0` and `kube-api-burst=100.0` such that the default settings make sense together with `udpate-worker-count=10`.

"**udpate**-worker-count"

voelzmo commented:
@jackfrancis @adrianmoisey @raywainman 
With the current implementation, the minimum amount of checkpoints written rather depends on the number of concurrent workers: Each worker picks a VPA to work on in parallel and at least one of them will pick a second VPA. So we have `minCheckpoints >= concurrentWorkers`. 
Should we remove this configuration setting entirely and mention this in the cli doc for `concurrentWorkers`? WDYT?

jackfrancis commented on vertical-pod-autoscaler/pkg/recommender/checkpoint/checkpoint_writer.go:
```diff
@@ -76,48 +77,94 @@ func getVpasToCheckpoint(clusterVpas map[model.VpaID]*model.Vpa) []*model.Vpa {
+			select {
+			case <-ctx.Done():
+				if minCheckpoints <= 0 {
+					klog.V(0).InfoS("Failed to store checkpoints", "err", ctx.Err())
```
In the current implementation (if I'm reading it correctly) we only consider <= 0 minCheckpoints an error if the ctx is also in an error case. Are we overloading things here and covering the error and non-error case with a single Info log message? (in which case we might not want to use the failure language)

voelzmo replied:
If I understand correctly, you're asking if this log statement can also be reached in a non-error case and if we then should re-phrase it?
```
klog.V(0).InfoS("Failed to store checkpoints", "err", ctx.Err())
```

My understanding is that the conditions for this log line to be reached are
* the outside context has been cancelled (meaning: not all checkpoints could be written within the recommender's configured checkpoint timeout interval, [which defaults to 1 minute](https://github.com/kubernetes/autoscaler/blob/27a9a16b80f155042750de9665c38502f22ea9a9/vertical-pod-autoscaler/pkg/recommender/routines/recommender.go#L37)). This is definitely an error that the operator should take care of
* the minimum amount of checkpoints have been written. In large-scale environments, it could happen that all previous steps inside the recommender (loading all pods, loading the metrics, updating all VPAs) already took longer than 1 minute and the context is already cancelled when entering the `StoreCheckpoints` method. To ensure a minimum amount of progress in this case, we don't exit immediately, but write at least some checkpoints before exiting.

If `minCheckpoints <= 0` and the context _hasn't_ been cancelled yet, processing should continue as usual.

Does this make sense?

While writing this, I thought how to make this a bit more explicit:
* as [proposed above](https://github.com/kubernetes/autoscaler/pull/7992#issuecomment-2788860880), we could remove the `minCheckpoint` parameter entirely and only rely on the `concurrentWorkers` configuration, as the amount of checkpoints which get written for an already cancelled context depends on this config setting. This would also simplify the code quite a bit.
* Change the log line to mention the configured checkpoint timeout as in `Failed to store all checkpoints within the configured 'checkpoints-timeout' of 1 minute`

WDYT?

jackfrancis replied:
I like your suggestions. I think my original observation was that in the current implementation (before this change) we aren't indicating failure when the context is cancelled

```go
		// Draining ctx.Done() channel. ctx.Err() will be checked if timeout occurred, but minCheckpoints have
		// to be written before return from this function.
		select {
		case <-ctx.Done():
		default:
		}
```

Your response suggests that this change is intentional, and an improvement. Makes sense.

I especially like the idea of ditching `minCheckpoint`, as it seems like a very implementation detail-y accounting property that can take some time to grok.

voelzmo replied:
Added a commit to remove the recommender flag, ensure that checkpoint progress is made even when the outside context is cancelled immediately, and added a new release note mentioning the breaking change.

jackfrancis replied:
lgtm!

jackfrancis commented:
/lgtm

adrianmoisey commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender.go:
```diff
@@ -35,7 +37,6 @@ import (
 
 var (
 	checkpointsWriteTimeout = flag.Duration("checkpoints-timeout", time.Minute, `Timeout for writing checkpoints since the start of the recommender's main loop`)
-	minCheckpointsPerRun    = flag.Int("min-checkpoints", 10, "Minimum number of checkpoints to write per recommender's main loop")
```
Just a thought, Kubernetes doesn't remove flags in a single release (or ever?), they are normally deprecated over time.

What if this flag was still active, but spat out a warning if used? May be its description also updated:
```
minCheckpointsPerRun    = flag.Int("min-checkpoints", 10, "Minimum number of checkpoints to write per recommender's main loop. DEPRECATED FLAG and will be removed in the future.")
```

Then in a later release it can be remove altogether?

voelzmo replied:
And by "active" you mean the recommender wouldn't crash if you specify it, but its contents are ignored? That's probably a good idea – not sure if people are reading their logs more thoroughly than the release notes, though ;) I understand this is a pattern in k8s that we could (should?) follow as well.

However, if possible, I'd rather not like to keep it actually working, given that we would be changing twice how the minimum amount of checkpoints written works and how  much more complex the code is compared to just removing this entirely.

Do you want me to keep the flag but make it a no-op, such that people can see a deprecation warning in their logs?

adrianmoisey replied:
> And by "active" you mean the recommender wouldn't crash if you specify it, but its contents are ignored?

Correct. 

> That's probably a good idea – not sure if people are reading their logs more thoroughly than the release notes, though ;) I understand this is a pattern in k8s that we could (should?) follow as well.

Agreed. Keeping the flag around (without it doing anything) for a few releases isn't that much better than disabling it outright, but in theory we could check if 3rd party helm charts have updated to stop supporting this flag.
 
> Do you want me to keep the flag but make it a no-op, such that people can see a deprecation warning in their logs?

Correct, my vote is that the flag still exists, but is a no-op, and spits out a deprecation warning to users.

voelzmo replied:
done.

tkukushkin commented:
Hi! Sorry for the off-topic question. Since you're going to remove `min-checkpoints`, I have a related question regarding `checkpoints-timeout` – could you please clarify its purpose?

I see that the recommender updates VPAs first (without any timeout), and then updates checkpoints using `checkpoints-timeout`.  
From what I understand, most people typically use a VPA for the main container in each deployment, so the number of VPAs and checkpoints is usually about the same. That means updating VPAs and checkpoints should take roughly the same amount of time.

If the number of workers is not properly configured and updating VPAs takes more than [five minutes](https://github.com/kubernetes/autoscaler/blob/bf23616cf09d8ba23ecc9cb27ab86490efef6d70/vertical-pod-autoscaler/pkg/recommender/main.go#L144), the recommender fails its health check. If it's deployed via the Helm chart, it's quite likely to include a liveness probe, which would cause the pod to be restarted. In that case, `checkpoints-timeout` has no effect at all.

I mean it feels a bit inconsistent that there's no timeout for updating VPAs, but there is one for checkpoints.

In my case, I have around 4k VPAs and 4k checkpoints. That's quite a lot. Updating the checkpoints takes about 20 seconds (with `kube-api-qps=200`). Given that, I'm struggling to come up with a scenario where having `checkpoints-timeout` would be useful.

Thank you in advance for your reply!

tkukushkin commented:
@voelzmo
> but if people are configuring e.g. 1 to keep the existing behavior, we're back to the current state.

I don't really understand who would do that, it makes little sense. But okay, in general, I understand your position and this question is not very important to me personally. Thanks again!

voelzmo commented:
@adrianmoisey @jackfrancis this is good-to-go from my point of view. I think I've addressed all the review comments. Could you give this another round of review, please?
The only change since the last round should be that the `min-checkpoints` flag remains on the recommender for now, so it doesn't break immediately all consumers (helm charts, users building their own Deployments, etc). We can remove that flag in the following release then.

adrianmoisey commented:
Generally this LGTM, there's just a small comment around the handling of the `min-checkpoints` flag, but otherwise happy to approve this

voelzmo commented on vertical-pod-autoscaler/pkg/recommender/routines/recommender.go:
```diff
@@ -35,9 +37,15 @@ import (
+func init() {
+	minCheckpointsPerRun := flag.Int("min-checkpoints", 10, "Minimum number of checkpoints to write per recommender's main loop. WARNING: this flag is deprecated and doesn't have any effect. It will be removed in a future release. Refer to update-worker-count to influence the minimum number of checkpoints written per loop.")
+	if *minCheckpointsPerRun != 10 { // Default value is 10
+		klog.Warningf("The 'min-checkpoints' flag is deprecated and has no effect. It will be removed in a future release.")
```
> I can't get the recommender to spit out the warning.

That's a good catch. Seems like log output from the `init` method isn't showing up there. Not sure where this belongs then.

omerap12 replied:
We migrated VPA to use structural logging, let's change it to `InfoS` please.
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md?plain=1#L267

voelzmo replied:
Thanks for the good docs, @omerap12! 

I see that it also covers one of the points we weren't sure about: how to adapt the `Fatal` log calls. Contrary to what we've been doing, it recommends to _not_ use `os.Exit`, but use https://pkg.go.dev/k8s.io/klog/v2#FlushAndExit

omerap12 replied:
Yup! I dunno how I missed it lol.
I will open an issue to address it

omerap12 commented:
/lgtm 
@voelzmo , can we unhold?

voelzmo commented:
/hold cancel

Thanks for the reviews!

adrianmoisey commented:
🎉 🎉 🎉
